File: server/vance-shared/src/main/java/de/mhus/vance/shared/document/kind/ApplicationCodec.java
package de.mhus.vance.shared.document.kind;

import java.util.LinkedHashMap;
import java.util.Map;
import org.jspecify.annotations.Nullable;
import tools.jackson.core.JacksonException;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

/**
 * Parser and serialiser for {@code kind: application} document
 * bodies. JSON and YAML; no Markdown (a Mermaid-style fence makes
 * no sense for an app manifest).
 *
 * <p>The header carries both {@code kind} and {@code app} as scalar
 * meta keys, so {@code DocumentDocument.headers.app} is populated
 * automatically through the standard
 * {@link de.mhus.vance.shared.document.JsonHeaderStrategy} /
 * {@link de.mhus.vance.shared.document.YamlHeaderStrategy} machinery.
 *
 * <p>Body parsing is deliberately structural — app-specific typing
 * (lanes, gantt config, …) lives in companion helpers like
 * {@code CalendarsAppConfig} so this codec stays domain-agnostic.
 */
public final class ApplicationCodec {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final TypeReference<LinkedHashMap<String, Object>> JSON_MAP =
            new TypeReference<>() {};

    private ApplicationCodec() {
        // utility class
    }

    public static ApplicationDocument parse(String body, @Nullable String mimeType) {
        if (isJson(mimeType)) return parseJson(body);
        if (isYaml(mimeType)) return parseYaml(body);
        throw new KindCodecException("Unsupported mime type for application: " + mimeType);
    }

    public static String serialize(ApplicationDocument doc, @Nullable String mimeType) {
        if (isJson(mimeType)) return serializeJson(doc);
        if (isYaml(mimeType)) return serializeYaml(doc);
        throw new KindCodecException("Unsupported mime type for application: " + mimeType);
    }

    public static boolean supports(@Nullable String mimeType) {
        return isJson(mimeType) || isYaml(mimeType);
    }

    private static boolean isJson(@Nullable String mime) {
        return "application/json".equals(mime);
    }

    private static boolean isYaml(@Nullable String mime) {
        return "application/yaml".equals(mime)
                || "application/x-yaml".equals(mime)
                || "text/yaml".equals(mime)
                || "text/x-yaml".equals(mime);
    }

    // ── JSON / YAML ────────────────────────────────────────────────

    private static ApplicationDocument parseJson(String body) {
        if (body.isBlank()) return ApplicationDocument.empty("");
        Map<String, Object> parsed;
        try {
            parsed = JSON.readValue(body, JSON_MAP);
        } catch (JacksonException e) {
            throw new KindCodecException("Invalid JSON: " + e.getOriginalMessage(), e);
        }
        if (parsed == null) throw new KindCodecException("Top-level JSON must be an object");
        return promoteToDocument(KindHeaderCodec.unwrapJsonMeta(parsed));
    }

    private static ApplicationDocument parseYaml(String body) {
        if (body.isBlank()) return ApplicationDocument.empty("");
        return promoteToDocument(KindHeaderCodec.parseYamlBody(body));
    }

    private static String serializeJson(ApplicationDocument doc) {
        Map<String, Object> headerExtra = new LinkedHashMap<>();
        if (doc.app() != null && !doc.app().isBlank()) headerExtra.put("app", doc.app());
        // wrapJsonMeta only takes kind + body, so we build the
        // wrapped form by inlining the scalar header extras into
        // the $meta block manually for JSON (mirrors what
        // dumpYamlBody(kind, body, headerExtra) does for YAML).
        Map<String, Object> meta = new LinkedHashMap<>();
        meta.put("kind", canonicalKind(doc));
        for (Map.Entry<String, Object> e : headerExtra.entrySet()) meta.put(e.getKey(), e.getValue());
        Map<String, Object> wrapped = new LinkedHashMap<>();
        wrapped.put(KindHeaderCodec.META_KEY, meta);
        wrapped.putAll(buildBody(doc));
        try {
            return JSON.writerWithDefaultPrettyPrinter().writeValueAsString(wrapped) + "\n";
        } catch (JacksonException e) {
            throw new KindCodecException("Failed to write JSON: " + e.getOriginalMessage(), e);
        }
    }

    private static String serializeYaml(ApplicationDocument doc) {
        Map<String, Object> headerExtra = new LinkedHashMap<>();
        if (doc.app() != null && !doc.app().isBlank()) headerExtra.put("app", doc.app());
        return KindHeaderCodec.dumpYamlBody(canonicalKind(doc), buildBody(doc), headerExtra);
    }

    // ── Promotion ──────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private static ApplicationDocument promoteToDocument(Map<String, Object> obj) {
        String kind = (obj.get("kind") instanceof String s) ? s : "";
        String app = (obj.get("app") instanceof String s2) ? s2 : "";
        String title = (obj.get("title") instanceof String s3 && !s3.isBlank()) ? s3 : null;
        String description = (obj.get("description") instanceof String s4 && !s4.isBlank()) ? s4 : null;

        // App-specific config lives nested under the app's own name —
        // e.g. config.calendar = { window, lanes, gantt, conflicts }.
        // We rebuild a top-level config map containing every key that
        // isn't part of the manifest's reserved scalar set, so future
        // app faces can sit beside calendar without schema work.
        Map<String, Object> config = new LinkedHashMap<>();
        Map<String, Object> extra = new LinkedHashMap<>();
        for (Map.Entry<String, Object> e : obj.entrySet()) {
            String key = e.getKey();
            if (isReservedTopKey(key)) continue;
            if (e.getValue() instanceof Map<?, ?>) {
                config.put(key, (Map<String, Object>) e.getValue());
            } else {
                extra.put(key, e.getValue());
            }
        }
        return new ApplicationDocument(
                kind.isEmpty() ? "application" : kind,
                app, title, description, config, extra);
    }

    private static boolean isReservedTopKey(String key) {
        return switch (key) {
            case "kind", "app", "title", "description" -> true;
            default -> false;
        };
    }

    // ── Body builder ───────────────────────────────────────────────

    private static Map<String, Object> buildBody(ApplicationDocument doc) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (doc.title() != null) body.put("title", doc.title());
        if (doc.description() != null) body.put("description", doc.description());
        // Restore the app-config blocks at the top level.
        for (Map.Entry<String, Object> e : doc.config().entrySet()) {
            if (!body.containsKey(e.getKey())) body.put(e.getKey(), e.getValue());
        }
        for (Map.Entry<String, Object> e : doc.extra().entrySet()) {
            if (!body.containsKey(e.getKey())) body.put(e.getKey(), e.getValue());
        }
        return body;
    }

    private static String canonicalKind(ApplicationDocument doc) {
        return (doc.kind() == null || doc.kind().isBlank()) ? "application" : doc.kind();
    }
}

File: client/packages/vance-face/src/document/applicationCodec.ts
// Codec for `kind: application` documents — the manifest at the root
// of a Vance "app folder" (à la macOS .app bundles). Carries the
// app-type discriminator ($meta.app) plus an app-specific nested
// config block (e.g. `calendar: { lanes, gantt, conflicts, window }`
// for app=calendar).
//
// JSON and YAML only — Markdown makes no sense for an app manifest.
//
// Mirrors the server-side `ApplicationCodec` in `vance-shared`.

import {
  dumpYamlBody,
  parseYamlBody,
  unwrapJsonMeta,
} from './documentHeaderCodec';

export interface ApplicationDocument {
  /** Always `'application'`. */
  kind: string;
  /** App-type discriminator. Known v1 value: `'calendar'`. */
  app: string;
  title?: string;
  description?: string;
  /** App-specific configuration, keyed by the app type at the top
   *  level (e.g. `config.calendar = { … }`). The nested form lets a
   *  folder host multiple app faces in v2 without schema work. */
  config: Record<string, unknown>;
  extra: Record<string, unknown>;
}

export class ApplicationCodecError extends Error {
  constructor(message: string, public override readonly cause?: unknown) {
    super(message);
    this.name = 'ApplicationCodecError';
  }
}

// ── MIME helpers ─────────────────────────────────────────────────────

function isJson(mime: string): boolean {
  return mime === 'application/json';
}
function isYaml(mime: string): boolean {
  return mime === 'application/yaml'
    || mime === 'application/x-yaml'
    || mime === 'text/yaml'
    || mime === 'text/x-yaml';
}

// ── Public API ───────────────────────────────────────────────────────

export function parseApplication(body: string, mimeType: string): ApplicationDocument {
  if (isJson(mimeType)) return parseApplicationJson(body);
  if (isYaml(mimeType)) return parseApplicationYaml(body);
  throw new ApplicationCodecError(`Unsupported mime type for application: ${mimeType}`);
}

export function serializeApplication(doc: ApplicationDocument, mimeType: string): string {
  if (isJson(mimeType)) return serializeApplicationJson(doc);
  if (isYaml(mimeType)) return serializeApplicationYaml(doc);
  throw new ApplicationCodecError(`Unsupported mime type for application: ${mimeType}`);
}

export function isApplicationMime(mimeType: string | null | undefined): boolean {
  if (!mimeType) return false;
  return isJson(mimeType) || isYaml(mimeType);
}

export function emptyApplication(app: string = ''): ApplicationDocument {
  return { kind: 'application', app, config: {}, extra: {} };
}

// ── JSON ─────────────────────────────────────────────────────────────

function parseApplicationJson(body: string): ApplicationDocument {
  if (body.trim() === '') return emptyApplication();
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch (e) {
    throw new ApplicationCodecError(
      'Invalid JSON: ' + (e instanceof Error ? e.message : String(e)),
      e,
    );
  }
  if (!isObject(parsed)) {
    throw new ApplicationCodecError('Top-level JSON must be an object');
  }
  return promoteToDocument(unwrapJsonMeta(parsed));
}

function serializeApplicationJson(doc: ApplicationDocument): string {
  // $meta needs both kind + app inline — we build the wrapper by
  // hand since wrapJsonMeta only knows about `kind`.
  const meta: Record<string, unknown> = { kind: doc.kind || 'application' };
  if (doc.app) meta.app = doc.app;
  const wrapped: Record<string, unknown> = {
    $meta: meta,
    ...buildBody(doc),
  };
  return JSON.stringify(wrapped, null, 2) + '\n';
}

// ── YAML ─────────────────────────────────────────────────────────────

function parseApplicationYaml(body: string): ApplicationDocument {
  if (body.trim() === '') return emptyApplication();
  let merged: Record<string, unknown>;
  try {
    merged = parseYamlBody(body);
  } catch (e) {
    throw new ApplicationCodecError(
      'Invalid YAML: ' + (e instanceof Error ? e.message : String(e)),
      e,
    );
  }
  return promoteToDocument(merged);
}

function serializeApplicationYaml(doc: ApplicationDocument): string {
  const headerExtra: Record<string, unknown> = {};
  if (doc.app) headerExtra.app = doc.app;
  return dumpYamlBody(doc.kind || 'application', buildBody(doc), headerExtra);
}

// ── Shared promotion ─────────────────────────────────────────────────

function promoteToDocument(obj: Record<string, unknown>): ApplicationDocument {
  const kind = typeof obj.kind === 'string' && obj.kind ? obj.kind : 'application';
  const app = typeof obj.app === 'string' ? obj.app : '';
  const title = typeof obj.title === 'string' && obj.title ? obj.title : undefined;
  const description = typeof obj.description === 'string' && obj.description
    ? obj.description : undefined;

  const config: Record<string, unknown> = {};
  const extra: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (k === 'kind' || k === 'app' || k === 'title' || k === 'description') continue;
    if (isObject(v)) {
      config[k] = v;
    } else {
      extra[k] = v;
    }
  }
  return { kind, app, title, description, config, extra };
}

function buildBody(doc: ApplicationDocument): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (doc.title) body.title = doc.title;
  if (doc.description) body.description = doc.description;
  for (const [k, v] of Object.entries(doc.config)) {
    if (!(k in body)) body[k] = v;
  }
  for (const [k, v] of Object.entries(doc.extra)) {
    if (!(k in body)) body[k] = v;
  }
  return body;
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

// ── App-type helpers ─────────────────────────────────────────────────

export function applicationAppType(doc: ApplicationDocument): string {
  return doc.app || '';
}
